python/ql/lib/semmle/python/regex.qll

```diff
@@ -446,6 +446,45 @@ abstract class RegexString extends Expr {
     )
   }
 
+  /**
+   * A sequence of 'normal' characters.
+   */
+  predicate normalCharacterSequence(int start, int end) {
+    this.normalCharacter(start, end) and
+    end = start + 1 and
+    exists(int x, int y | this.charSet(x, y) and x <= start and y >= end)
+    or
+    exists(int s, int e |
+      e = max(int i | normalCharacterSub(s, i)) and
+      not exists(int x, int y | this.charSet(x, y) and x <= s and y >= e)
```

Comment:
Having recently fixed some unfortunate uses of `not exists` in this file, I was a bit wary of this construction. I think it's probably fine, though, since we're materialising the `charSet` relation anyway.